File: 300-lambda.rb
TWO  = ->(f) { ->(v) { f.(f.(v)) } }

# WAT.....
DECREMENT = ->(n) { ->(f) { ->(x) {
  n.(->(g) { ->(h) { h.(g.(f)) } }).(->(y) { x }).(->(y){ y })
} } }


def to_integer(num)
  num.(->(x) { x + 1}).(0)
end

puts to_integer(DECREMENT.(TWO))

# next - multiply

File: js/300-lambda.js
const TWO  = (f) => { return (v) => { return f(f(v)) }}

// WAT.....
const DECREMENT = (n) => {
  return (f) => { return (x) => {
    return n((g) => { return (h) => { return h(g(f))}})((y) => { return x})((y) => { return y })
  }}
}


function toInt(x) {
  return x((v) => { return v + 1 })(0)
}

console.log(toInt(DECREMENT(TWO)))

// next - multiply
